readme completed

fn1z opened the pull request.

Commit:
readme completed

mateacademy-ai-mentor commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,27 @@
+            return false;
+        }
+
+        boolean ageValid = candidate.getAge() >= 35;
```
This violates the checklist item: "Use constants 
Magic numbers and strings decrease code readability, let's avoid it and use constants in class CandidateValidator" — the literal 35 is used here for minimum eligible age. Introduce a named constant (e.g. MIN_ELIGIBLE_AGE = 35) and use it instead.

mateacademy-ai-mentor commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,27 @@
+
+        boolean ageValid = candidate.getAge() >= 35;
+        boolean allowed = candidate.isAllowedToVote();
+        boolean ukrainian = "Ukrainian".equalsIgnoreCase(candidate.getNationality());
```
This violates the checklist item: "Use constants 
Magic numbers and strings decrease code readability, let's avoid it and use constants in class CandidateValidator" — the literal string "Ukrainian" is used here. Move this value to a named constant (e.g. REQUIRED_NATIONALITY = "Ukrainian") and use it in the comparison.

mateacademy-ai-mentor commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,27 @@
+        }
+        int from = Integer.parseInt(years[0].trim());
+        int to = Integer.parseInt(years[1].trim());
+        boolean lived10Years = (to - from) >= 10;
```
This violates the checklist item: "Use constants 
Magic numbers and strings decrease code readability, let's avoid it and use constants in class CandidateValidator" — the literal 10 is used here for minimum years lived in Ukraine. Introduce a named constant (e.g. MIN_YEARS_IN_UKRAINE = 10) and use it instead.

mateacademy-ai-mentor commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -14,7 +18,13 @@ public class StreamPractice {
+                .mapToInt(Integer::parseInt)
+                .filter(n -> n % 2 == 0)
+                .min()
+                .orElseThrow(() ->
+                        new RuntimeException("Can't get min value from list: " + numbers));
```
This uses OptionalInt.orElseThrow with a Supplier (the lambda). The primitive Optional types (OptionalInt) may not support orElseThrow(Supplier) in some JDK versions and this can cause a compilation error. The checklist suggests handling absent results via primitive getters or appropriate terminal ops: "If you need to throw exception when no result found after stream execution sometimes `get()/getAsDouble()` may help, try to google what does it do." 

Recommendation: box before min (e.g. `.mapToInt(...).boxed().min(...).orElseThrow(() -> new RuntimeException(...))`) so you get an Optional<Integer> that accepts a Supplier, or use OptionalInt.orElseThrow() no-arg and then rethrow/wrap to produce the exact RuntimeException message required.

mateacademy-ai-mentor commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -23,7 +33,11 @@ public int findMinEvenNumber(List<String> numbers) {
+                .map(i -> (i % 2 == 1) ? numbers.get(i) - 1 : numbers.get(i))
+                .filter(n -> n % 2 != 0)
+                .average()
+                .orElseThrow(NoSuchElementException::new);
```
average() returns OptionalDouble; calling orElseThrow with a Supplier (NoSuchElementException::new) on OptionalDouble may not compile because primitive optionals differ from Optional<T>. The checklist hints: "If you need to throw exception when no result found after stream execution sometimes `get()/getAsDouble()` may help, try to google what does it do." 

Recommendation: use `.average().orElseThrow()` (no-arg) which throws NoSuchElementException, or change the method to return primitive `double` and use `.average().getAsDouble()` so that a NoSuchElementException is thrown when empty. This will align with the requirement to throw NoSuchElementException when no odd numbers remain.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision:** REQUEST_CHANGES — there are a few critical checklist violations and potential compilation blockers that need fixing.

**Critical issues to fix (blockers):**
1. CandidateValidator.java — magic literals used directly (checklist: “Use constants”).
   - Where: age check uses `35` (e.g., around the age comparison line), nationality comparison uses `"Ukrainian"`, and years-in-Ukraine check uses `10`.
   - Why: Explicit checklist item; using magic numbers/strings is not allowed.
   - How: Introduce constants and use them in logic, e.g. `private static final int MIN_ELIGIBLE_AGE = 35;`, `private static final String REQUIRED_NATIONALITY = "Ukrainian";`, `private static final int MIN_YEARS_IN_UKRAINE = 10;`.

2. StreamPractice.java — use of Supplier overload on primitive optionals may not compile.
   - findMinEvenNumber: `.mapToInt(...).min().orElseThrow(() -> new RuntimeException(...))` uses `OptionalInt.orElseThrow(Supplier)` which may not be available in the target JDK.
     - How: Either box before `min` and use `Optional<Integer>`: `.mapToInt(...).boxed().min(...).orElseThrow(() -> new RuntimeException("Can't get min value from list: " + numbers))`, or use `OptionalInt` safely: check `isPresent()` and `getAsInt()`, then throw the `RuntimeException` with the exact message when empty.
   - getOddNumsAverage: `.average().orElseThrow(NoSuchElementException::new)` on `OptionalDouble` may not compile.
     - How: Use `.average().orElseThrow()` (no-arg) and keep returning `Double`, or call `.average().getAsDouble()` (it throws `NoSuchElementException` when empty). Ensure the thrown exception is `NoSuchElementException` per requirement.

**Minor/optional improvements (non-blockers):**
- CandidateValidator: make parsing of `periodsInUkr` more robust (null/blank check; split with `"\\s*-\\s*"`; handle NumberFormatException and return false).
- Javadoc wording: if comments say “older than 35,” align text with logic “at least 35”.

Your stream logic and overall approach are solid; after introducing constants and adjusting the two Optional usages, the solution should compile cleanly and fully meet the checklist. Keep it up!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

Commit:
mateacademy-ai-mentor fixes

Commit:
fixed

mateacademy-ai-mentor commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,41 @@
+
+        boolean ageValid = candidate.getAge() >= MIN_ELIGIBLE_AGE;
+        boolean allowed = candidate.isAllowedToVote();
+        boolean ukrainian = REQUIRED_NATIONALITY.equalsIgnoreCase(candidate.getNationality());
```
This uses case-insensitive comparison for nationality. The task requires nationality to equal "Ukrainian" (case-sensitive). Change `REQUIRED_NATIONALITY.equalsIgnoreCase(candidate.getNationality())` to a case-sensitive comparison, e.g. `REQUIRED_NATIONALITY.equals(candidate.getNationality())`. (Requirement: "have nationality - 'Ukrainian'")

mateacademy-ai-mentor commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,41 @@
+        String[] years = periods.split("\\s*-\\s*"); // split with optional spaces
+        if (years.length != 2) {
+            return false;
+        }
```
Parsing here splits the entire field by hyphen and requires exactly two tokens, which fails when `periodsInUkr` contains multiple ranges (for example "1999-2001,2004-2014"). The validator must support multiple periods and sum their durations. Quote from requirements: "lived in Ukraine for at least 10 years: parse periodsInUkr entries like '2002-2015', sum inclusive differences (end - start), and ensure total >= 10 years. If multiple periods exist, sum all of them." Update logic to first split by comma into ranges, then split each range by hyphen and sum the parsed durations.

mateacademy-ai-mentor commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,41 @@
+            int from = Integer.parseInt(years[0]);
+            int to = Integer.parseInt(years[1]);
+            boolean lived10Years = (to - from) >= MIN_YEARS_IN_UKRAINE;
+            return ageValid && allowed && ukrainian && lived10Years;
```
Current try-block handles only one `from`/`to`. Replace it with logic that iterates (or streams) over each parsed range, parses start/end for each, accumulates the durations, and then checks whether the total meets `MIN_YEARS_IN_UKRAINE`. Also consider whitespace robustness and how to handle malformed ranges (skip or treat as 0).

mateacademy-ai-mentor commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -23,7 +35,11 @@ public int findMinEvenNumber(List<String> numbers) {
+                .map(i -> (i % 2 == 1) ? numbers.get(i) - 1 : numbers.get(i))
+                .filter(n -> n % 2 != 0)
+                .average()
+                .orElseThrow();
```
`getOddNumsAverage` ends with `.average().orElseThrow()`. To ensure the method throws `NoSuchElementException` on empty and to avoid potential JDK compatibility issues with `orElseThrow()`, replace this with `.average().getAsDouble()`. The primitive result will be autoboxed to `Double` for the method return type. This aligns with the checklist hint to use `get()/getAsDouble()` when you want an exception thrown by an empty Optional/OptionalDouble. See this call in StreamPractice.java.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Educational approval: moving this forward so you can keep progressing, while highlighting a few fixes to strengthen your solution.

What you did well
- Clear use of constants in `CandidateValidator` and clean Predicate design. Stream tasks are mostly correct and readable, with proper enum comparisons and stream chaining.

Critical issues to address
1) CandidateValidator.java
   - Line ~20: Nationality check uses `equalsIgnoreCase`, but the requirement is exact match to "Ukrainian". Why: the spec says "have nationality - 'Ukrainian'" (case-sensitive). Hint: use `REQUIRED_NATIONALITY.equals(candidate.getNationality())`.
   - Lines ~27–36: `periodsInUkr` parsing handles only a single `YYYY-YYYY`. Why: inputs like `"1999-2001,2004-2014"` must be supported and summed. Hint: split by commas first (e.g., `periods.split("\\s*,\\s*")`), then for each range split by hyphen, parse `from`/`to`, accumulate `(to - from)` (clarify inclusive/exclusive per your spec), and compare the total to `MIN_YEARS_IN_UKRAINE`. Be robust to whitespace and malformed ranges (skip or treat as 0; keep your try/catch).
   - Minor: If targeting Java 8, replace `isBlank()` with `trim().isEmpty()`.

2) StreamPractice.java
   - In `getOddNumsAverage`: replace `.average().orElseThrow()` with `.average().getAsDouble()` to throw `NoSuchElementException` as per the checklist and avoid JDK compatibility issues.

Keep going—your structure is solid, and these targeted tweaks will bring everything in line with the requirements.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

Elena-Bruyako commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,41 @@
+import java.util.function.Predicate;
+import model.Candidate;
+
+public class CandidateValidator implements Predicate<Candidate> { // minor comment
```
remove comment

Elena-Bruyako commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,41 @@
+
+    @Override
+    public boolean test(Candidate candidate) {
+        if (candidate == null) {
```
let's make it simplier
```suggestion
        return candidate.getAge() >= MIN_AGE
                && candidate.isAllowedToVote()
                && candidate.getNationality().equals(REQUIRED_NATIONALITY)
                && checkTimeLivingInCountry(candidate);
```

create method 
```
private boolean checkTimeLivingInCountry(Candidate candidate) {
        String[] dates = candidate.getPeriodsInUkr().split(SEPARATOR);
        return Integer.parseInt(dates[END_INDEX])
                - Integer.parseInt(dates[BEGIN_INDEX]) >= PERIOD_IN_UKR;
    }
```

Elena-Bruyako commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -13,8 +17,16 @@ public class StreamPractice {
-        return 0;
+    public int findMinEvenNumber(List<String> numbers) { // minor comment
+        return numbers.stream()
+                .flatMap(s -> Arrays.stream(s.split("\\s*,\\s*")))
```
```suggestion
                .flatMap(s -> Arrays.stream(input.split(",")))
```

Elena-Bruyako commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -13,8 +17,16 @@ public class StreamPractice {
+                .mapToInt(Integer::parseInt)
+                .filter(n -> n % 2 == 0)
+                .boxed()
+                .min(Comparator.naturalOrder())
```
```suggestion
                .map(Integer::parseInt)
                .filter(num -> num % 2 == 0)
                .min(Integer::compareTo)
```

Elena-Bruyako commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -35,7 +51,10 @@ public Double getOddNumsAverage(List<Integer> numbers) {
+        return peopleList.stream()
+                .filter(p -> p.getSex() == Person.Sex.MAN)
+                .filter(p -> p.getAge() >= fromAge && p.getAge() <= toAge)
+                .collect(Collectors.toList());
```
```suggestion
                .toList();
```

Elena-Bruyako commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -50,7 +69,11 @@ public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toA
+                .filter(p -> p.getSex() == Person.Sex.MAN
+                        ? (p.getAge() >= fromAge && p.getAge() <= maleToAge)
+                        : (p.getAge() >= fromAge && p.getAge() <= femaleToAge))
+                .collect(Collectors.toList());
```
```suggestion
                .toList();
```

Elena-Bruyako commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -59,11 +82,16 @@ public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
+                .filter(p -> p.getAge() >= femaleAge)
+                .flatMap(p -> p.getCats().stream())
+                .map(Cat::getName)
+                .collect(Collectors.toList());
```
```suggestion
                .toList();
```

Elena-Bruyako commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -75,6 +103,11 @@ public List<String> getCatsNames(List<Person> peopleList, int femaleAge) {
+                .filter(validator)
+                .map(Candidate::getName)
+                .sorted()
+                .collect(Collectors.toList());
```
```suggestion
                .toList();
```

Elena-Bruyako replied:
fix comment

Elena-Bruyako reviewed (changes requested):
see comments

Commit:
removed comments
simplified CandidateValidator
findMinEvenNumber method fixed
simplified .collect(Collectors.toList()); to .toList();

Elena-Bruyako reviewed (changes requested):
one comment

Commit:
all .collect(Collectors.toList()); -> .toList();

Elena-Bruyako reviewed (approved):
well done
